pkg/prometheus/agent/operator.go

```diff
@@ -544,13 +568,6 @@ func statefulSetKeyToPrometheusAgentKey(key string) (bool, string) {
 
 // Sync implements the operator.Syncer interface.
 func (c *Operator) Sync(ctx context.Context, key string) error {
-	err := c.sync(ctx, key)
-	c.reconciliations.SetStatus(key, err)
-
-	return err
-}
-
-func (c *Operator) sync(ctx context.Context, key string) error {
```

Comment:
I would keep the `sync()` method and only branch when the operator updates the statefulsets. What do you think?

Reply:
You mean keeping the creating logic in sync() for both DaemonSet and StatefulSet? But their creating logics are not entirely similar. For example, for StatefulSet we need to check storageclass and handle sharding before creating.

In the end, we can surely refactor the common operator logics between two modes, whether in creating, updating or deleting. But maybe we can do that later, when we already have a working implmentation of DaemonSet?

For #6652 I understand we needed to do some refactoring before it, because it would have been harder to review and verify without the refactoring. But for operator's logic we can easily verify with local tests and e2e tests, so I don't think refactoring is a priority at the moment?

Reply:
I assume that we can still optimize a bit (e.g. reduce duplication). In particular everything which isn't related to statefulset/daemonset should be common?

Reply:
Yeah we can still do it. Tbh operator's logic is still somewhat cryptic for me, so I just want to focus on making it work for DaemonSet first. But if there're any parts that seems obvious and easy to refactor, I'll do it then^^

Reply:
Now took a deeper look into `syncDaemonSet()` and `syncStatefulSet()`, I think we have refactored as far as we can. Yes there's still some common codes left. But these common codes do different things, so I'm not sure if we shoudl refactor them further into one function.

@ArthurSens Could you check and see if you agree with this?

@simonpasquier Even if we decide to merge without refactoring further now, you can always take another review when you come back from vacation^^

Reply:
Yeah, this code block is a duplication. But it does different things here. Should we really refactor them into _one_ function?

Maybe one function `beforeCreatingAgentClient()`? :thinking:

Reply:
@simonpasquier wdyt^^

Reply:
Let's create an issue to revisit potential refactoring after this PR merges.

Reply:
I didn't create an issue. Instead I created a TODO. All the things we wish to do later have been added a TODO in this PR. Will that work too?